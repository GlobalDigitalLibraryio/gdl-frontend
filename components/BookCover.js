// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */
import * as React from 'react';
import { css } from 'react-emotion';

import CoverImage from './CoverImage';
import type { Book } from '../types';
import theme from '../style/theme';

type Props = {
  book: Book
};

const style = css`
  position: relative;
  box-shadow: ${theme.boxShadows.small};
`;

const BookCover = ({ book }: Props) => (
  <CoverImage
    className={style}
    src={book.coverPhoto && book.coverPhoto.large}
    h="100%"
    w="100%"
  />
);

export default BookCover;
