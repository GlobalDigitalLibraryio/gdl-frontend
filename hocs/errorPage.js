// @flow
import * as React from 'react';
import ErrorPage from '../pages/_error';
import type { Context } from '../types';

export default (Page: React.ComponentType<*>) =>
  class extends React.Component<*> {
    static async getInitialProps(ctx: Context) {
      // $FlowFixMe This is okay, we're only wrapping other pages
      const props = await Page.getInitialProps(ctx);

      // On the server, get the status code from the response
      const resStatusCode = ctx.res && ctx.res.statusCode;

      // Get the statusCode from the wrapped page
      const { statusCode } = props;

      // If we're on the server, and the response status code doesn't match that of the wrapped component, we set it
      if (ctx.res && statusCode && ctx.res.statusCode !== statusCode) {
        ctx.res.statusCode = statusCode;
      }

      return { statusCode: resStatusCode, ...props };
    }

    render() {
      const { statusCode } = this.props;
      if (statusCode && statusCode !== 200) {
        return <ErrorPage statusCode={statusCode} />;
      }
      return <Page {...this.props} />;
    }
  };
