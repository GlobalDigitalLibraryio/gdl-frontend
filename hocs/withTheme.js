// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 * 
 * See LICENSE
 */

import * as React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';

/* eslint-disable react/no-multi-comp */

const theme = {
  primaries: {
    primary: '#20588f',
    secondary: '#507aa4',
    tertiary: '#a5bcd3',
    light: '#ceddea',
    highlight: '#edf7ff',
    dark: '#184673',
  },
  supports: {
    greenPrimary: '#5cbc80',
    greenHighlight: '#edfff4',
    greenDark: '#359258',
  },
  grays: {
    dark: '#444',
    white: '#fff',
    steel: '#666',
    jumbo: '#888',
    silverSand: '#bbb',
    platinum: '#e3e3e3',
    gallery: '#eff0f2',
    desertStorm: '#f8f8f8',
  },
};

class Theme extends React.Component<{ children: React.Node }> {
  static childContextTypes = {
    reactIconBase: PropTypes.object,
  };

  getChildContext() {
    return {
      reactIconBase: {
        size: 24,
      },
    };
  }

  render() {
    return <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>;
  }
}

/**
 * HoC that wraps a page with Theme
 */
function withTheme(Page: React.ComponentType<any>) {
  return class PageWithTheme extends React.Component<any> {
    static getInitialProps(context) {
      if (typeof Page.getInitialProps === 'function') {
        return Page.getInitialProps(context);
      }
      return undefined;
    }

    render() {
      return (
        <Theme>
          <Page {...this.props} />
        </Theme>
      );
    }
  };
}

export { Theme, theme, withTheme as default };
