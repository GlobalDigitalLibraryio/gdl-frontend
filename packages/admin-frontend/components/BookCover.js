// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */
import * as React from 'react';
import { css } from 'react-emotion';

import CoverImage from './CoverImage';
import type { CoverImageInfo } from '../types';
import misc from "../style/misc"

type Props = {
  coverImage: ?CoverImageInfo,
  w: Array<string | number>,
  h: Array<string | number>
};

const style = css`
  position: relative;
  box-shadow: ${misc.boxShadows.small};
`;

const BookCover = ({ coverImage, w, h }: Props) => (
  <CoverImage
    ariaHidden
    className={style}
    src={coverImage && coverImage.url}
    alt={coverImage && coverImage.alttext}
    h={h}
    w={w}
  />
);

export default BookCover;
