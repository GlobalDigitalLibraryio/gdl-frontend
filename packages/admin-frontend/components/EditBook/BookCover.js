// @flow
import * as React from 'react';
import { ImageOutlined as ImageIcon } from '@material-ui/icons';
import { imageUrl } from 'gdl-image';
import { Button, Typography } from '@material-ui/core';
import { fetchImageMetadata } from '../../lib/fetch';
import type { BookDetails, ImageMetadata } from '../../types';
import EditBookCoverDialog from './EditBookCoverDialog';

type Props = {
  book: BookDetails
};

type State = {
  dialogOpen: boolean,
  imageMetadata: ?ImageMetadata
};

export default class EditBookImage extends React.Component<Props, State> {
  state = {
    dialogOpen: false,
    imageMetadata: null
  };

  componentDidMount() {
    this.getImageData();
  }

  async getImageData() {
    if (!this.props.book.coverImage) return;

    const result = await fetchImageMetadata(this.props.book.coverImage.imageId);

    if (result.isOk) {
      this.setState({ imageMetadata: result.data });
    }
  }

  handleSave = () => {
    this.getImageData();
    this.setState({ dialogOpen: false });
  };

  handleCancel = () => this.setState({ dialogOpen: false });

  render() {
    if (!this.props.book.coverImage) {
      return null;
    }

    const { imageMetadata } = this.state;

    // While we're waiting for the imageApi to respond with the metadata, we use the stuff we know from the book object
    const altText = imageMetadata
      ? imageMetadata.alttext.alttext
      : this.props.book.coverImage.alttext;

    const coverImage = imageMetadata
      ? {
          url: imageMetadata.imageUrl,
          imageId: imageMetadata.id,
          variants: imageMetadata.imageVariants,
          alttext: altText
        }
      : this.props.book.coverImage;

    return (
      <>
        <div css={{ textAlign: 'center' }}>
          <img
            src={imageUrl(coverImage, { aspectRatio: 0.81 })}
            css={{
              width: 310,
              height: 380,
              objectFit: 'cover',
              objectPosition: 'center center'
            }}
            alt={altText}
          />
          <Typography variant="caption" gutterBottom>
            Alt: {altText && <em>{altText}</em>}
          </Typography>

          <Button
            onClick={() => this.setState({ dialogOpen: true })}
            size="small"
          >
            <ImageIcon /> Edit book cover
          </Button>
        </div>
        {this.state.dialogOpen &&
          imageMetadata && (
            <EditBookCoverDialog
              onCancel={this.handleCancel}
              onSave={this.handleSave}
              imageMetadata={imageMetadata}
            />
          )}
      </>
    );
  }
}
