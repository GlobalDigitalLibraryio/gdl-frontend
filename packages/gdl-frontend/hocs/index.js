// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2018 GDL
 *
 * See LICENSE
 */

export { default as withErrorPage } from './withErrorPage';
export { default as securePage } from './securePage';
