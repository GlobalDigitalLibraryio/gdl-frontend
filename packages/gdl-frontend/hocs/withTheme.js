// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */

import * as React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'emotion-theming';
import { MuiThemeProvider } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import JssProvider from 'react-jss/lib/JssProvider';

import getPageContext from '../getPageContext';
import type { Context } from '../types';
import { misc } from '../style/theme';

const theme = { space: [], breakpoints: misc.breakpoints };

class Theme extends React.Component<{ children: React.Node }> {
  static childContextTypes = {
    reactIconBase: PropTypes.object
  };

  getChildContext() {
    return {
      reactIconBase: {
        size: misc.iconSize
      }
    };
  }

  render() {
    return <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>;
  }
}

//const jss = create(jssPreset());

/**
 * HoC that wraps a page with Theme
 */
function withTheme(Page: React.ComponentType<any>) {
  return class PageWithTheme extends React.Component<any> {
    static getInitialProps(context: Context) {
      if (typeof Page.getInitialProps === 'function') {
        return Page.getInitialProps(context);
      }
      return undefined;
    }

    constructor(props, context) {
      super(props, context);
      this.pageContext = this.props.pageContext || getPageContext();
    }

    componentDidMount() {
      // Remove the server-side injected CSS.
      const jssStyles = document.querySelector('#jss-server-side');
      if (jssStyles && jssStyles.parentNode) {
        jssStyles.parentNode.removeChild(jssStyles);
      }
    }

    render() {
      return (
        <Theme>
          <JssProvider
            jss={this.pageContext.jss}
            registry={this.pageContext.sheetsRegistry}
            generateClassName={this.pageContext.generateClassName}
          >
            <MuiThemeProvider sheetsManager={this.pageContext.sheetsManager}>
              <CssBaseline />
              <Page {...this.props} />
            </MuiThemeProvider>
          </JssProvider>
        </Theme>
      );
    }
  };
}

export { Theme, withTheme as default };
