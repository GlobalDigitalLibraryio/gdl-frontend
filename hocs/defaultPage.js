// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */

import * as React from 'react';
import Router from 'next/router';
import withI18n from './withI18n';
import withTheme from './withTheme';
import withErrorBoundary from './withErrorBoundary';
import {
  setAnonToken,
  getAccessTokenFromRequest,
  getAccessTokenFromLocalStorage,
  setAnonTokenOnResponse,
  LOGOUT_KEY,
} from '../lib/auth/token';
import { fetchAnonToken } from '../fetch';
import { logPageView } from '../lib/analytics';


if (typeof window !== 'undefined') {
  logPageView();
}

/**
 * HoC that combines all necessary pages wrapper so we get a single point of entry
 * It also makes sure we have a token to the APIs
 */

const defaultPage = Page =>
  class DefaultPage extends React.Component<any> {
    static async getInitialProps(ctx) {
      const accessToken = process.browser
        ? getAccessTokenFromLocalStorage()
        : getAccessTokenFromRequest(ctx.req);

      let fullToken;
      // If there is no access token and we're on the server, generate one and set it as a cookie
      if (!accessToken) {
        fullToken = await fetchAnonToken();
        // On the client, this is set in componentDidMount()
        if (!process.browser) {
          setAnonTokenOnResponse(ctx.res, fullToken);
        }
      }

      ctx.accessToken = accessToken || (fullToken && fullToken.access_token);

      // Evaluate the composed component's getInitialProps()
      let composedInitialProps;
      // Check if it actually is a next page
      if (typeof Page.getInitialProps === 'function') {
        composedInitialProps = await Page.getInitialProps(ctx);
      }

      return {
        fullToken,
        ...composedInitialProps,
      };
    }

    componentDidMount() {
      // cDM only runs on the client, so if we get a token via props, set it in local storage.
      // We cannot read it from the cookies on the client, because we won't be able to access the expiration date then.
      if (this.props.fullToken) {
        setAnonToken(this.props.fullToken);
      }
      window.addEventListener('storage', this.logout, false);
    }

    componentWillUnmount() {
      window.removeEventListener('storage', this.logout, false);
    }

    // $FlowFixMe: StorageEvent was added to Flow in v0.58.0, so remove this line after we've upgraded 
    logout = (event: StorageEvent) => {
      if (event.key === LOGOUT_KEY) {
        Router.push(`/?logout=${event.newValue}`);
      }
    }

    render() {
      const { token, ...props } = this.props;
      return <Page {...props} />;
    }
  };

export default (
  Page: React.ComponentType<any>,
  wrapWithErrorBoundary: boolean = true,
) =>
  defaultPage(
    withTheme(withI18n(wrapWithErrorBoundary ? withErrorBoundary(Page) : Page)),
  );
