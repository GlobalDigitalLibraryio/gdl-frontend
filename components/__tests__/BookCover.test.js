// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */

import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';
import type { Book } from '../../types';
import { theme } from '../../hocs/withTheme';
import BookCover from '../BookCover';

const book: Book = {
  id: 1,
  title: 'A Life in the Fair Jungle',
  description:
    'Sindi buzzes with colorful songs and dance. Will starting school end all that?',
  license: {
    id: 1,
    name: 'cc by 4.0',
    description: 'Attribution 4.0 International (CC BY 4.0)',
    url: 'https://creativecommons.org/licenses/by/4.0/',
  },
  language: {
    code: 'eng',
    name: 'English',
  },
  categories: [],
  contributors: [],
  readingLevel: '3',
  availableLanguages: [],
  publisher: {
    id: 1,
    name: 'Pratham Books',
  },
  coverPhoto: {
    large:
      'http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg',
    small:
      'http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg?width=200',
  },
  chapters: [],
  downloads: {
    epub: '',
    pdf: '',
  },
};

test('Renders', () => {
  const tree = mount(<BookCover book={book} theme={theme} />);

  expect(toJson(tree)).toMatchSnapshot();
});
