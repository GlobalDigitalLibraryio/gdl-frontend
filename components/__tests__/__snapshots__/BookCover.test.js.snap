// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
.c1 {
  max-height: 100%;
  max-width: 100%;
}

.c0 {
  width: 150px;
  height: 185px;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0.12),0 2px 2px 0 rgba(0,0,0,0.12);
  border-radius: 4px;
  color: #444;
  background-color: #f8f8f8;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media screen and (min-width:40em) {
  .c0 {
    width: 200px;
  }
}

@media screen and (min-width:40em) {
  .c0 {
    height: 225px;
  }
}

<BookCover
  book={
    Object {
      "availableLanguages": Array [],
      "categories": Array [],
      "chapters": Array [],
      "contributors": Array [],
      "coverPhoto": Object {
        "large": "http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg",
        "small": "http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg?width=200",
      },
      "description": "Sindi buzzes with colorful songs and dance. Will starting school end all that?",
      "downloads": Object {
        "epub": "",
        "pdf": "",
      },
      "id": 1,
      "language": Object {
        "code": "eng",
        "name": "English",
      },
      "license": Object {
        "description": "Attribution 4.0 International (CC BY 4.0)",
        "id": 1,
        "name": "cc by 4.0",
        "url": "https://creativecommons.org/licenses/by/4.0/",
      },
      "publisher": Object {
        "id": 1,
        "name": "Pratham Books",
      },
      "readingLevel": "3",
      "title": "A Life in the Fair Jungle",
    }
  }
  h={
    Array [
      "185px",
      "225px",
    ]
  }
  theme={
    Object {
      "grays": Object {
        "dark": "#444",
        "desertStorm": "#f8f8f8",
        "gainsboro": "#ddd",
        "gallery": "#eff0f2",
        "jumbo": "#888",
        "platinum": "#e3e3e3",
        "steel": "#666",
        "white": "#fff",
      },
      "primaries": Object {
        "dark": "#184673",
        "highlight": "#edf7ff",
        "light": "#ceddea",
        "primary": "#20588f",
        "secondary": "#507aa4",
        "tertiary": "#a5bcd3",
      },
      "supports": Object {
        "greenDark": "#359258",
        "greenHighlight": "#edfff4",
        "greenPrimary": "#5cbc80",
      },
    }
  }
  w={
    Array [
      "150px",
      "200px",
    ]
  }
>
  <styled.div
    h={
      Array [
        "185px",
        "225px",
      ]
    }
    theme={
      Object {
        "grays": Object {
          "dark": "#444",
          "desertStorm": "#f8f8f8",
          "gainsboro": "#ddd",
          "gallery": "#eff0f2",
          "jumbo": "#888",
          "platinum": "#e3e3e3",
          "steel": "#666",
          "white": "#fff",
        },
        "primaries": Object {
          "dark": "#184673",
          "highlight": "#edf7ff",
          "light": "#ceddea",
          "primary": "#20588f",
          "secondary": "#507aa4",
          "tertiary": "#a5bcd3",
        },
        "supports": Object {
          "greenDark": "#359258",
          "greenHighlight": "#edfff4",
          "greenPrimary": "#5cbc80",
        },
      }
    }
    w={
      Array [
        "150px",
        "200px",
      ]
    }
  >
    <div
      className="c0"
    >
      <Component
        alt="A Life in the Fair Jungle"
        src="http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg?width=200"
      >
        <styled.img
          alt="A Life in the Fair Jungle"
          aria-hidden={true}
          src="http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg?width=200"
        >
          <img
            alt="A Life in the Fair Jungle"
            aria-hidden={true}
            className="c1"
            src="http://test-proxy-1865761686.eu-central-1.elb.amazonaws.com/image-api/v1/raw/2-smile-please-image_1.jpg?width=200"
          />
        </styled.img>
      </Component>
    </div>
  </styled.div>
</BookCover>
`;
